### ISSUE TITLE
Missing error handling leading to undefined behaviour

### ISSUE THREAD

**Issue by Flamefire** [ComputationalRadiationPhysics/mallocMC#110]
The `MALLOCMC_CUDA_CHECKED_CALL` macro uses `CUDA::checkError` to check for error codes which throws an error in case of non-zero values. But this happens only in Debug mode.

Therefore when the common `SimpleCudaMalloc` policy is used and `setMemPool` is called with an invalid/to high size or the `cudaMalloc` call fails for any reason, an uninitialized pointer is returned.
This also holds true for e.g. sizes of 0 in which case cudaMalloc does not modify the pointer but returns a zero-error code indicating success. See the code:

    static void* setMemPool(size_t memsize){
      void* pool;
      MALLOCMC_CUDA_CHECKED_CALL(cudaMalloc(&pool, memsize));
      return pool;
    }

Even setting this pointer to zero and not throwing an error would not solve the problem, as the result of the call is never checked and implicitly assumed valid in all following usages.

As the minimum action to counter this I propose to enable the exception throwing also in Release mode and initialize the pointer to NULL before passing it to cudaMalloc

**Comment by slizzered** [comment]
I implemented your suggestions in #112. 

This still does not solve the problem where the size is `0`. Since it is an allowed standard behaviour of malloc (and by extension probably cudaMalloc) succeed in this case without an error, it should probably be treated as acceptable behaviour here. As you pointed out in #111, checking for NULL would be a good idea inside the CreationPolicy itself as necessary (like Scatter).

**Comment by Flamefire** [comment]
I think this should suffice. For the case of size=0 a NULL pointer and a 0 size is returned which should make the Scatter Policy create 0 pages. So this should be fine. Only thing to check, if the scatter policy expects non-zero sizes and creates e.g. a counter always. But from a quick glance it should be fine.

**Comment by ax3l** [comment]
can this issue be closed for a next release?
The PIConGPU release depends on it since it introduces mallocMC.

**Comment by Flamefire** [comment]
IMO the policy must still check for a NULL pointer. This is currently not the case. However as a NULL pointer means that zero memory was requested (otherwise an error would have been thrown) this should not cause problems.
However at least #112 must be merged before a release.

**Comment by ax3l** [comment]
@Flamefire can you provide a PR for @slizzered ?

**Comment by Flamefire** [comment]
For the NULL pointer check? The other issue will be fixed with #112?

**Comment by ax3l** [comment]
#112 says it "fixes this issue (#110)" - is this true or does it need a follow up / your mentioned NULL pointer check?

**Comment by Flamefire** [comment]
Considering only the description of this issue #112 would close this as it will work correctly with the current implementation. However I'll open a follow-up issue for the missing NULL pointer check which will only be a problem, if another allocation policy is used, that does not throw an error on a failed malloc but returns a NULL pointer.

**Comment by ax3l** [comment]
thank you, feel free to link this issue for context.

**Comment by Flamefire** [comment]
Just seen that I already mentioned this in #111.
As explained there the NULL pointer check in the scatter policy is a must before a release. Otherwise one could combine the reserve-only policy with the scatter policy leading to a NULL pointer passed undetected to the scatter policy and used till a crash terminates the program (possibly not... Had cases where device side memory errors lead to hanging programs on hypnos)

**Comment by ax3l** [comment]
ok that is why it is lost because it is mixed in #111 and here, but not an independent issue.
